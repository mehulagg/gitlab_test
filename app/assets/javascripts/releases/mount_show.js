import Vue from 'vue';
import ReleaseShowApp from './components/app_show.vue';
import createStore from './stores';
import createDetailModule from './stores/modules/detail';

export default () => {
  const el = document.getElementById('js-show-release-page');

  const store = createStore({
    modules: {
      detail: createDetailModule(el.dataset),
    },
  });

  return new Vue({
    el,
    store,
    render: h => h(ReleaseShowApp),
  });
};
