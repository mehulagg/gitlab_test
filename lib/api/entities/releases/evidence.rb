# frozen_string_literal: true

module API
  module Entities
    module Releases
      class Evidence < Grape::Entity
        expose :summary_sha, as: :sha
        expose :filepath
        expose :collected_at

        def filepath
          release = object.release
          project = release.project

          Gitlab::Routing.url_helpers.namespace_project_release_evidence_url(namespace_id: project.namespace_id, project_id: project, release_tag: release, id: object.id, format: :json)
        end
      end
    end
  end
end