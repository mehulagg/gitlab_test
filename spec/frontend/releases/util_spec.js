import { cloneDeep } from 'lodash';
import { releaseToApiJson, apiJsonToRelease, convertGraphQLResponse } from '~/releases/util';
import { graphqlReleasesResponse as originalGraphqlReleasesResponse } from './mock_data';

describe('releases/util.js', () => {
  describe('releaseToApiJson', () => {
    it('converts a release JavaScript object into JSON that the Release API can accept', () => {
      const release = {
        tagName: 'tag-name',
        name: 'Release name',
        description: 'Release description',
        milestones: [{ id: 1, title: '13.2' }, { id: 2, title: '13.3' }],
        assets: {
          links: [{ url: 'https://gitlab.example.com/link', linkType: 'other' }],
        },
      };

      const expectedJson = {
        tag_name: 'tag-name',
        ref: null,
        name: 'Release name',
        description: 'Release description',
        milestones: ['13.2', '13.3'],
        assets: {
          links: [{ url: 'https://gitlab.example.com/link', link_type: 'other' }],
        },
      };

      expect(releaseToApiJson(release)).toEqual(expectedJson);
    });

    describe('when createFrom is provided', () => {
      it('adds the provided createFrom ref to the JSON as a "ref" property', () => {
        const createFrom = 'main';

        const release = {};

        const expectedJson = {
          ref: createFrom,
        };

        expect(releaseToApiJson(release, createFrom)).toMatchObject(expectedJson);
      });
    });

    describe('release.name', () => {
      it.each`
        input                 | output
        ${null}               | ${null}
        ${''}                 | ${null}
        ${' \t\n\r\n'}        | ${null}
        ${'  Release name  '} | ${'Release name'}
      `('converts a name like `$input` to `$output`', ({ input, output }) => {
        const release = { name: input };

        const expectedJson = {
          name: output,
        };

        expect(releaseToApiJson(release)).toMatchObject(expectedJson);
      });
    });

    describe('when release.milestones is falsy', () => {
      it('includes a "milestone" property in the returned result as an empty array', () => {
        const release = {};

        const expectedJson = {
          milestones: [],
        };

        expect(releaseToApiJson(release)).toMatchObject(expectedJson);
      });
    });
  });

  describe('apiJsonToRelease', () => {
    it('converts JSON received from the Release API into an object usable by the Vue application', () => {
      const json = {
        tag_name: 'tag-name',
        assets: {
          links: [
            {
              link_type: 'other',
            },
          ],
        },
      };

      const expectedRelease = {
        tagName: 'tag-name',
        assets: {
          links: [
            {
              linkType: 'other',
            },
          ],
        },
        milestones: [],
      };

      expect(apiJsonToRelease(json)).toEqual(expectedRelease);
    });
  });

  describe('convertGraphQLResponse', () => {
    let graphqlReleasesResponse;
    let converted;

    beforeEach(() => {
      graphqlReleasesResponse = cloneDeep(originalGraphqlReleasesResponse);
      converted = convertGraphQLResponse(graphqlReleasesResponse);
    });

    it('matches snapshot', () => {
      expect(converted).toMatchSnapshot();
    });

    describe('assets', () => {
      it("handles asset links that don't have a linkType", () => {
        expect(converted.data[0].assets.links[0].linkType).not.toBeUndefined();

        delete graphqlReleasesResponse.data.project.releases.nodes[0].assets.links.nodes[0]
          .linkType;

        converted = convertGraphQLResponse(graphqlReleasesResponse);

        expect(converted.data[0].assets.links[0].linkType).toBeUndefined();
      });
    });

    describe('_links', () => {
      it("handles releases that don't have any links", () => {
        expect(converted.data[0]._links.selfUrl).not.toBeUndefined();

        delete graphqlReleasesResponse.data.project.releases.nodes[0].links;

        converted = convertGraphQLResponse(graphqlReleasesResponse);

        expect(converted.data[0]._links.selfUrl).toBeUndefined();
      });
    });

    describe('commit', () => {
      it("handles releases that don't have any commit info", () => {
        expect(converted.data[0].commit).not.toBeUndefined();

        delete graphqlReleasesResponse.data.project.releases.nodes[0].commit;

        converted = convertGraphQLResponse(graphqlReleasesResponse);

        expect(converted.data[0].commit).toBeUndefined();
      });
    });
  });
});
