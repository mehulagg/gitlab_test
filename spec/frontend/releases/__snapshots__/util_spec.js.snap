// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`releases/util.js convertGraphQLResponse matches snapshot 1`] = `
Object {
  "data": Array [
    Object {
      "_links": Object {
        "editUrl": "http://0.0.0.0:3000/root/release-test/-/releases/v5.10/edit",
        "issuesUrl": null,
        "mergeRequestsUrl": null,
        "self": "http://0.0.0.0:3000/root/release-test/-/releases/v5.10",
        "selfUrl": "http://0.0.0.0:3000/root/release-test/-/releases/v5.10",
      },
      "assets": Object {
        "count": 7,
        "links": Array [
          Object {
            "directAssetUrl": "http://0.0.0.0:3000/root/release-test/-/releases/v5.32/permanent/path/to/runbook",
            "external": true,
            "id": "gid://gitlab/Releases::Link/69",
            "linkType": "other",
            "name": "An example link",
            "url": "https://example.com/link",
          },
          Object {
            "directAssetUrl": "https://example.com/package",
            "external": true,
            "id": "gid://gitlab/Releases::Link/68",
            "linkType": "package",
            "name": "An example package link",
            "url": "https://example.com/package",
          },
          Object {
            "directAssetUrl": "https://example.com/image",
            "external": true,
            "id": "gid://gitlab/Releases::Link/67",
            "linkType": "image",
            "name": "An example image",
            "url": "https://example.com/image",
          },
        ],
        "sources": Array [
          Object {
            "format": "zip",
            "url": "http://0.0.0.0:3000/root/release-test/-/archive/v5.10/release-test-v5.10.zip",
          },
          Object {
            "format": "tar.gz",
            "url": "http://0.0.0.0:3000/root/release-test/-/archive/v5.10/release-test-v5.10.tar.gz",
          },
          Object {
            "format": "tar.bz2",
            "url": "http://0.0.0.0:3000/root/release-test/-/archive/v5.10/release-test-v5.10.tar.bz2",
          },
          Object {
            "format": "tar",
            "url": "http://0.0.0.0:3000/root/release-test/-/archive/v5.10/release-test-v5.10.tar",
          },
        ],
      },
      "author": Object {
        "avatarUrl": "/uploads/-/system/user/avatar/1/avatar.png",
        "username": "root",
        "webUrl": "http://0.0.0.0:3000/root",
      },
      "commit": Object {
        "shortId": "92e7ea2e",
        "title": "Testing a change.",
      },
      "commitPath": "http://0.0.0.0:3000/root/release-test/-/commit/92e7ea2ee4496fe0d00ff69830ba0564d3d1e5a7",
      "descriptionHtml": "<p data-sourcepos=\\"1:1-1:24\\" dir=\\"auto\\">This is version <strong>1.0</strong>!</p>",
      "evidences": Array [
        Object {
          "collectedAt": "2020-08-21T20:15:19Z",
          "filepath": "http://0.0.0.0:3000/root/release-test/-/releases/v5.10/evidences/34.json",
          "sha": "22bde8e8b93d870a29ddc339287a1fbb598f45d1396d",
        },
      ],
      "milestones": Array [
        Object {
          "description": "",
          "id": "gid://gitlab/Milestone/60",
          "issueStats": Object {
            "closed": 0,
            "total": 0,
          },
          "stats": undefined,
          "title": "12.4",
          "webPath": undefined,
          "webUrl": "/root/release-test/-/milestones/2",
        },
        Object {
          "description": "Milestone 12.3",
          "id": "gid://gitlab/Milestone/59",
          "issueStats": Object {
            "closed": 1,
            "total": 2,
          },
          "stats": undefined,
          "title": "12.3",
          "webPath": undefined,
          "webUrl": "/root/release-test/-/milestones/1",
        },
      ],
      "name": "Release 1.0",
      "releasedAt": "2020-08-21T20:15:18Z",
      "tagName": "v5.10",
      "tagPath": "/root/release-test/-/tags/v5.10",
      "upcomingRelease": false,
    },
  ],
  "paginationInfo": Object {
    "endCursor": "eyJpZCI6IjMiLCJyZWxlYXNlZF9hdCI6IjIwMjAtMDctMDkgMjA6MTE6MzMuODA0OTYxMDAwIFVUQyJ9",
    "hasNextPage": true,
    "hasPreviousPage": false,
    "startCursor": "eyJpZCI6IjQ0IiwicmVsZWFzZWRfYXQiOiIyMDMwLTAzLTE1IDA4OjAwOjAwLjAwMDAwMDAwMCBVVEMifQ",
  },
}
`;
